File: targets/js/sample-counter-v3/src/mvu/state-holder.ts
export class StateHolder<TState> {
  private _state: TState;

  onChange: (() => void) | null = null;

  constructor(initial: TState) {
    this._state = initial;
  }

  get state(): TState {
    return this._state;
  }

  set(next: TState): void {
    this._state = next;
    this.onChange?.();
  }
}

File: samples/SampleCounterV2/Mvu/StateHolder.cs
namespace SampleCounterV2.Mvu;

public sealed class StateHolder<TState>(TState initial)
{
    private TState _state = initial;

    public TState State => _state;

    public Action? OnChange { get; set; }

    public void Set(TState next)
    {
        _state = next;
        OnChange?.Invoke();
    }
}
